metadata-ingestion/src/datahub/ingestion/source/iceberg/iceberg.py

```diff
@@ -313,16 +313,30 @@ def _parse_datatype(type: IcebergTypes.Type, nullable: bool = False) -> Dict[str
             "_nullable": nullable,
         }
     elif type.is_map_type():
+        # The Iceberg Map type will be handled differently.  The idea is to translate the map
+        # similar to the Map.Entry struct of Java i.e. as an array of map_entry struct, where
+        # the map_entry struct has a key field and a value field. The key and value type can
+        # be complex or primitive types.
         map_type: IcebergTypes.MapType = type
-        kt = _parse_datatype(map_type.key_type())
-        vt = _parse_datatype(map_type.value_type())
-        # keys are assumed to be strings in avro map
+        map_entry: Dict[str, Any] = {
+            "type": "record",
+            "name": _gen_name("__map_entry_"),
+            "fields": [
+                {
+                    "name": "key",
+                    "type": _parse_datatype(map_type.key_type(), False),
+                },
+                {
+                    "name": "value",
+                    "type": _parse_datatype(map_type.value_type(), True),
+                },
+            ],
+        }
         return {
-            "type": "map",
-            "values": vt,
-            "native_data_type": str(map_type),
-            "key_type": kt,
-            "key_native_data_type": repr(map_type.key_type()),
+            "type": "array",
+            "items": map_entry,
```

Comment:
What ultimately happens to this `items` field? It is not part of SchemaField.pdl, so is it transformed into another field later?

Reply:
To be honest, I don't know.  In the original Iceberg source PR, it got reviewed by a member of the Iceberg core team and one thing he pointed out was that Iceberg supports complex key types for map.  Here is the [original comment](https://github.com/datahub-project/datahub/pull/4431#discussion_r871663832) (and you can read the following comments from me and Ravindra).  I only implemented what was suggested to me, so I'm not sure what was the original plan for the UI.

Reply:
Hi @gabe-lyons, we have a few non pure AVRO schema types(hive, iceberg) that are leveraging our existing AVRO to SchemaField converter implementation. To get around the key-type limitation that is restricted to only the string type in AVRO for the other schema types, we will need to special-case this mechanism (array[__map_entry]) also as a map type in the UI soon. We can discuss more offline.